powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/callbacks/CallbackUrlEventService.java

```diff
@@ -0,0 +1,336 @@
+            callbackUrlEventEntity.setTimestampDeleteAfter(callbackUrlEventEntity.getTimestampCreated().plus(retentionPeriod));
+        }
+
+        callbackUrlEventEntity.setStatus(CallbackUrlEventStatus.FAILED);
+        callbackUrlEventRepository.save(callbackUrlEventEntity);
+    }
+
+    /**
+     * Set a Callback URL Event as currently being processed.
+     * @param callbackUrlEventEntity The Callback URL Event to set as currently being processed.
+     */
+    private void markAsProcessing(final CallbackUrlEventEntity callbackUrlEventEntity) {
+        callbackUrlEventEntity.setStatus(CallbackUrlEventStatus.PROCESSING);
+        callbackUrlEventEntity.setTimestampNextCall(null);
+        callbackUrlEventEntity.setTimestampLastCall(LocalDateTime.now());
+        callbackUrlEventEntity.setAttempts(callbackUrlEventEntity.getAttempts() + 1);
+        callbackUrlEventRepository.save(callbackUrlEventEntity);
+    }
+
+    /**
+     * Calculate back off period for next retry attempt using exponential backoff strategy.
+     * @param attempts Number of already made attempts.
+     * @param initialBackoff Initial backoff.
+     * @return Duration between last and next attempt.
+     */
+     private static Duration calculateExponentialBackoffPeriod(final int attempts, final long initialBackoff, final double multiplier, final long maxBackoffMilliseconds) {
+        if (attempts < 0) {
+            throw new IllegalArgumentException("Attempts must be non-negative.");
+        }
+
+        if (initialBackoff < 0) {
+            throw new IllegalArgumentException("Initial backoff must be non-negative.");
+        }
+
+        if (attempts == 0) {
+            return Duration.ZERO;
+        }
+
+        final long backoff = (long) (initialBackoff * Math.pow(multiplier, attempts - 1));
+        return Duration.ofMillis(Math.min(backoff, maxBackoffMilliseconds));
+    }
+
+    /**
+     * Get a rest client for a callback url entity.
+     * @param callbackUrlEntity Callback url entity.
+     * @return Rest client.
+     * @throws RestClientException Thrown when rest client initialization fails.
+     */
+    private RestClient getRestClient(final CallbackUrlEntity callbackUrlEntity) throws RestClientException {
+        final String cacheKey = getRestClientCacheKey(callbackUrlEntity);
+        RestClient restClient;
+        synchronized (restClientCacheLock) {
+            restClient = restClientCache.get(cacheKey);
+            if (restClient == null) {
+                logger.debug("REST client not found in cache, initializing new REST client, callback cache key: {}", cacheKey);
+                restClient = createRestClientAndStoreInCache(callbackUrlEntity);
+            } else {
+                logger.debug("REST client found in cache, callback cache key: {}", cacheKey);
+            }
+        }
+        return restClient;
```

Comment:
Let's use _final_ modifier as much as possible.
```suggestion
        synchronized (restClientCacheLock) {
            final RestClient restClient = restClientCache.get(cacheKey);
            if (restClient == null) {
                logger.debug("REST client not found in cache, initializing new REST client, callback cache key: {}", cacheKey);
                return createRestClientAndStoreInCache(callbackUrlEntity);
            } else {
                logger.debug("REST client found in cache, callback cache key: {}", cacheKey);
                return restClient;
            }
        }
```